pyupgrade/_plugins/typing_pep563.py

```diff
@@ -0,0 +1,90 @@
+import ast
+import functools
+from typing import Iterable
+from typing import MutableMapping
+from typing import Optional
+from typing import Tuple
+from typing import Union
+
+from tokenize_rt import Offset
+
+from pyupgrade._ast_helpers import ast_to_offset
+from pyupgrade._data import register
+from pyupgrade._data import State
+from pyupgrade._data import TokenFunc
+from pyupgrade._token_helpers import dequote
+
+
+def _supported_version(state: State) -> bool:
+    return (
+        state.settings.min_version >= (3, 10) or (
+            'annotations' in state.from_imports['__future__']
+        )
+    )
+
+
+def replace_string_literal(
+    annotation: Optional[Union[ast.expr, ast.slice]],
+    to_replace: MutableMapping[ast.expr, str],
+) -> None:
+    if isinstance(annotation, ast.Constant):
+        if isinstance(annotation.value, str):  # pragma: no cover
+            # Not for Python 3.6, Python 3.7
+            to_replace[annotation] = annotation.value
+    elif isinstance(annotation, ast.Str):  # pragma: no cover
+        # Only for Python 3.6, Python 3.7
+        to_replace[annotation] = annotation.s
+    elif isinstance(annotation, (ast.Tuple, ast.List)):
+        for i in annotation.elts:
+            replace_string_literal(i, to_replace)
+    elif isinstance(annotation, ast.Index):  # pragma: no cover
+        # Not for Python 3.9
+        replace_string_literal(annotation.value, to_replace)
+    elif isinstance(annotation, ast.Subscript):
+        if isinstance(annotation.value, ast.Name):
+            if annotation.value.id != 'Literal':
+                replace_string_literal(annotation.slice, to_replace)
+
+
+@register(ast.FunctionDef)
+def visit_FunctionDef(
+    state: State,
+    node: ast.FunctionDef,
+    parent: ast.AST,
+) -> Iterable[Tuple[Offset, TokenFunc]]:
+    if not _supported_version(state):
+        return
+    to_replace: MutableMapping[ast.expr, str] = {}
+    for i in node.args.args:
+        replace_string_literal(i.annotation, to_replace)
```

Comment:
the rest of the code tries to avoid recursion here as it can easily bump into the limit on deeply nested constructs (this was one of the motivations for the plugin-based system)